src/sentry/profiles/task.py

```diff
@@ -618,35 +619,47 @@ def _deobfuscate(profile: Profile, project: Project) -> None:
 
     with sentry_sdk.start_span(op="proguard.remap"):
         for method in profile["profile"]["methods"]:
+            method.setdefault("data", {})
+
             mapped = mapper.remap_frame(
                 method["class_name"], method["name"], method["source_line"] or 0
             )
-            method.setdefault("data", {})
-            if len(mapped) == 1:
-                new_frame = mapped[0]
-                method.update(
-                    {
-                        "class_name": new_frame.class_name,
-                        "name": new_frame.method,
-                        "source_file": new_frame.file,
-                        "source_line": new_frame.line,
-                    }
-                )
-                method["data"]["deobfuscation_status"] = "deobfuscated"
-            elif len(mapped) > 1:
+
+            if "signature" in method and method["signature"]:
+                method["signature"] = deobfuscate_signature(mapper, method["signature"])
+
+            if len(mapped) >= 1:
+                new_frame = mapped[-1]
+                new_frame_attributes = {
+                    "class_name": new_frame.class_name,
+                    "name": new_frame.method,
+                    "data": {"deobfuscation_status": "deobfuscated"},
+                }
+
+                if new_frame.file:
+                    new_frame_attributes["source_file"] = new_frame.file
+
+                if new_frame.line:
+                    new_frame_attributes["source_line"] = new_frame.line
+
+                method.update(new_frame_attributes)
```

Comment:
Nit: feels like you should just update each attribute instead of conditionally adding them to a new dict and calling update but no big deal.

```python
method[...] = ...
```